multinet/multinet/__init__.py

```diff
@@ -93,6 +95,13 @@ def bulk(self, params, workspace=None, table=None):
         rows = csv.DictReader(StringIO(cherrypy.request.body.read().decode('utf8')))
         workspace = db.db(workspace)
 
+        # Check for key uniqueness
+        if ('_key' in rows.fieldnames):
+            keys = [row['_key'] for row in rows]
+            uniqueKeys = set(keys)
+            if len(keys) != len(uniqueKeys):
+                raise RestException('CSV Validation Failed')
```

Comment:
Can you augment this message with something like "non-unique key detected"?

Reply:
Sure thing. I could also specify which key was duplicated (or in the case of multiple, the first duplicate found). However, this would add a little more time complexity. I'll leave it to your judgement whether that's a real concern and/or if the functionality is worth it.

Reply:
That's a good idea.

I don't think the time complexity should increase though. You could just insert each key one by one into the set, and for each key that is already found there, record it as a duplicate key, then report them all at the end.

Insertion and lookup into a set should both be O(log n), and you're doing this with n values, for a total complexity of O(n log n). Currently, you're inserting all the keys into a set, which seems like an O(n log n) operation as well.

If we do go this route, we can just merge the current PR as-is and open a new issue describing this change.

Let me know what you think.

Reply:
Since Python implements sets as hash tables, I would think insertion and deletion would be O(1) (amortized). So creating a set of n elements should be O(n). Though I agree the way you outlined would have the same time complexity. I'm good with merging this and opening a new PR for that change.

Reply:
You're right, Python uses hash tables to implement sets. Learned something new today :)

Ok, since the situation is even better than O(n log n), let's just augment the message the way I requested, and then get this merged. I'll open an issue for the enhancement. Thanks @AlmightyYakob!

Reply:
Sounds good!